### ISSUE TITLE
No real speedup observed in the implementation

### ISSUE THREAD

**Issue by x-zho14** [YOUSIKI/PyTorch-FBS#1]
Hi, I have read your code and I find that there is no real speedup in the implementation. The output is calculated firstly and then multiply the mask. Can you please provide the code with speedup claimed in your rebuttal, https://openreview.net/forum?id=BJxh2j0qYm? I have implemented the channel selection procedure with masked_select operation while this costs much more than a vanilla convolution. Thank you!

**Comment by yousiki** [comment]
So happy that this repo is found by you.
Well, in this repo I simply implemented the method in this paper fully in PyTorch, 
which means that there is no actual speedup operations (which might need CUDA C support).
I created this repo several months ago, because I was trying to implement another unpublished idea based on this,
in which we examine the speed by flop ops (we havent test on REAL HARDWARE with fully support yet).
I thought this repo may be helpful for people who might also want to train FBS in pytorch, so I made it public.
However, if you want actual speedup, please refer to the author's official implementation which use BMXNet-v2 as deep-leanring framework.
Thanks your interaction. Hope my words can be helpful for your questions, and wish you success.

**Comment by x-zho14** [comment]
Thanks for your rapid reply! Can you please provide the link to the "author's official implementation" ? Thank you!

**Comment by yousiki** [comment]
Sorry that I misremembered. This is FBS's implementation:
https://github.com/deep-fry/mayo/blob/master/docs/fbs.md
BMXNet-v2 is a framework for weight quantization, and this repo's framework is mayo.
Forgive me for my bad memory.

**Comment by x-zho14** [comment]
Do this implementation of FBS(https://github.com/deep-fry/mayo/blob/master/docs/fbs.md) provides speedup? Or it is just another implementation based on mayo?

**Comment by yousiki** [comment]
I didn't try mayo but I remembered that it said that it HAS "hardware acceleration"?
Moreover, I just comed up with the idea that, maybe implementing speedup version in PyTorch is not that difficult? One can generate a new weight Tensor and a new input Tensor before calling torch.nn.functional.conv2d and reconstruct the real Tensor after passing that Conv2d layer. I believe you can further implement this in PyTorch with little lines of code. 
Best wishes.

**Comment by x-zho14** [comment]
Yeah, I have stated this implementation in the my first comment. Actually the construction process of new weight Tensor takes *much more time* than convolution and the reduction in channel cannot make up for it. So I believe there is still doubt on whether FBS leads to actual speedup claimed in your rebuttal.
